File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.js
/**
 * @param {number[][]} heights
 * @return {number[][]}
 */
var pacificAtlantic = function(heights) {
    const [m, n] = [heights.length, heights[0].length];
    const pacSet = new Set();
    const atlSet = new Set();
    let visited = new Set();
    const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    
    for (let r = 0; r < m; r++) {
        pacSet.add(JSON.stringify([r, 0]));
        atlSet.add(JSON.stringify([r, n - 1]));
    }
    
    for (let c = 0; c < n; c++) {
        pacSet.add(JSON.stringify([0, c]));
        atlSet.add(JSON.stringify([m - 1, c]));
    }
    
    const inBounds = (r, c) => {
        return r >= 0 && r < m && c >= 0 && c < n;
    };
    
    const valid = (sr, sc, dr, dc) => {
        return !visited.has(JSON.stringify([dr, dc])) && inBounds(dr, dc) && heights[sr][sc] <= heights[dr][dc];
    };
    
    const dfs = (r, c, ocean) => {
        if (!visited.has(JSON.stringify([r, c]))) {
            visited.add(JSON.stringify([r, c]));
            dirs.forEach(([rd, cd]) => {
                const [newR, newC] = [r + rd, c + cd];
                if (valid(r, c, newR, newC)) {
                    switch (ocean) {
                        case 'pac':
                            if (pacSet.has(JSON.stringify([r, c]))) pacSet.add(JSON.stringify([newR, newC]));
                            break;
                        case 'atl':
                            if (atlSet.has(JSON.stringify([r, c]))) atlSet.add(JSON.stringify([newR, newC]));
                            break;
                        default:
                            break;
                    }
                    dfs(newR, newC, ocean);
                }
            });
        }
    };
    
    Array.from(pacSet).map(el => JSON.parse(el)).forEach(([r, c]) => dfs(r, c, 'pac'));

    visited = new Set();
    
    Array.from(atlSet).map(el => JSON.parse(el)).forEach(([r, c]) => dfs(r, c, 'atl'));
    
    let intersect = new Set([...pacSet].filter(el => atlSet.has(el)));
    return Array.from(intersect).map(coord => JSON.parse(coord));
};

File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        m, n = len(heights), len(heights[0])
        pacSet = set()
        atlSet = set()
        visited = set()
        dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        
        for r in range(m):
            pacSet.add((r, 0))
            atlSet.add((r, n-1))
            
        for c in range(n):
            pacSet.add((0, c))
            atlSet.add((m-1, c))
            
        def inBounds(r, c):
            return r in range(m) and c in range(n)
        
        def valid(sr, sc, dr, dc):
            return (dr, dc) not in visited and inBounds(dr, dc) and heights[sr][sc] <= heights[dr][dc]
        
        def dfs(r, c, ocean):
            if (r, c) not in visited:
                visited.add((r, c))
                for rd, cd in dirs:
                    newR, newC = r+rd, c+cd
                    if valid(r, c, newR, newC):
                        if ocean == 'pac' and (r, c) in pacSet:
                            pacSet.add((newR, newC))
                        elif ocean == 'atl' and (r, c) in atlSet:
                            atlSet.add((newR, newC))
                        dfs(newR, newC, ocean)
                        
        for r, c in list(pacSet):
            dfs(r, c, 'pac')
            
        visited = set()
        
        for r, c in list(atlSet):
            dfs(r, c, 'atl')
            
        intersect = pacSet & atlSet
        return list(intersect)